crypto/dso/dso_dlfcn.c

```diff
@@ -308,6 +308,77 @@ static int dladdr(void *address, Dl_info *dl)
 }
 # endif                         /* __sgi */
 
+# ifdef _AIX
+/*-
+ * See IBM's AIX Version 7.2, Technical Reference:
+ *  Base Operating System and Extensions, Volume 1 and 2
+ *  https://www.ibm.com/support/knowledgecenter/ssw_aix_72/com.ibm.aix.base/technicalreferences.htm
+ */
+#  include <sys/ldr.h>
+#  include <errno.h>
+/* ~ 64 * (sizeof(struct ld_info) + _XOPEN_PATH_MAX + _XOPEN_NAME_MAX) */
+#  define DLFCN_LDINFO_SIZE 86976
+typedef struct Dl_info {
+    const char *dli_fname;
+} Dl_info;
+/*
+ * This dladdr()-implementation will also find the ptrgl (Pointer Glue) virtual
+ * address of a function, which is just located in the DATA segment instead of
+ * the TEXT segment.
+ */
+static int dladdr(void *addr, Dl_info *dl)
+{
+    unsigned int found = 0;
+    struct ld_info *ldinfos, *next_ldi, *this_ldi;
+
+    if ((ldinfos = (struct ld_info *)OPENSSL_malloc(DLFCN_LDINFO_SIZE)) == NULL) {
+        errno = ENOMEM;
+        dl->dli_fname = NULL;
+        return 0;
+    }
+
+    if ((loadquery(L_GETINFO, (void *)ldinfos, DLFCN_LDINFO_SIZE)) < 0) {
+        /*-
+         * Error handling is done through errno and dlerror() reading errno:
+         *  ENOMEM (ldinfos buffer is too small),
+         *  EINVAL (invalid flags),
+         *  EFAULT (invalid ldinfos ptr)
+         */
+        OPENSSL_free((void *)ldinfos);
+        dl->dli_fname = NULL;
+        return 0;
+    }
+    next_ldi = ldinfos;
+
+    do {
+        this_ldi = next_ldi;
+        if (((addr >= this_ldi->ldinfo_textorg)
+             && ((uintptr_t)addr <
+                 ((uintptr_t)this_ldi->ldinfo_textorg +
+                     this_ldi->ldinfo_textsize)))
+            || ((addr >= this_ldi->ldinfo_dataorg)
+                && ((uintptr_t)addr <
+                    ((uintptr_t)this_ldi->ldinfo_dataorg +
+                        this_ldi->ldinfo_datasize)))) {
+            found = 1;
+            /*
+             * Ignoring the possibility of a member name and just returning
+             * the path name. See docs: sys/ldr.h, loadquery() and
+             * dlopen()/RTLD_MEMBER.
+             */
+            if ((dl->dli_fname =
+                 OPENSSL_strdup(this_ldi->ldinfo_filename)) == NULL)
+                errno = ENOMEM;
+        } else {
+            next_ldi =
+                (struct ld_info *)((uintptr_t)this_ldi + this_ldi->ldinfo_next);
```

Comment:
For reference, in *this* line pointer arithmetic is actually compliant with standard, because result of addition still falls inside same object, one allocated in the beginning of this subroutine. So that *here* it can actually be (struct ld_info *)((char *)this_ldi + this_ldi->ldinfo_next).

Reply:
Is this about that warning you mentioned, @makr?

> It however does not like adding an unsigned long to a pointer.

Reply:
Your suggestion leads to:
```
"crypto/dso/dso_dlfcn.c", line 374.36: 1506-495 (I) Pointer type conversion found.
"crypto/dso/dso_dlfcn.c", line 374.36: 1506-374 (I) Pointer types "char*" and "struct ld_info*" are not compatible.
"crypto/dso/dso_dlfcn.c", line 374.17: 1506-495 (I) Pointer type conversion found.
"crypto/dso/dso_dlfcn.c", line 374.17: 1506-374 (I) Pointer types "struct ld_info*" and "char*" are not compatible.
```
So I stick with the uintptr_t cast here, too. Commit with the other changes follows...

Reply:
@levitte before changing to the uintptr_t cast it actually was a warning:
`"crypto/dso/dso_dlfcn.c", line 370.22: 1506-068 (W) Operation between types "struct ld_info*" and "char*" is not allowed.`

Reply:
I'll venture a guess that the compiler warns about that mix because that could lead to alignment faults, while it *has* to allow any calculation when `uintptr_t` used.  That would make some kind of sense.  Anyway, this indicates that a `uintptr_t` cast is the most correct way indeed.